src/library/scala/math/BigInt.scala

```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
+  }
 
   /** Returns a pair of two BigInts containing (this / that) and (this % that).
    */
   def /% (that: BigInt): (BigInt, BigInt) = {
+    if (this.longEncoding && that.longEncoding) {
+      val x = this._long
+      val y = that._long
+      if (x != Long.MinValue || y != -1) return (new BigInt(x / y), new BigInt(x % y))
+    }
     val dr = this.bigInteger.divideAndRemainder(that.bigInteger)
     (new BigInt(dr(0)), new BigInt(dr(1)))
   }
 
   /** Leftshift of BigInt
    */
-  def << (n: Int): BigInt = new BigInt(this.bigInteger.shiftLeft(n))
+  def << (n: Int): BigInt =
+    if (longEncoding && n <= 0) this >> (-n) else new BigInt(this.bigInteger.shiftLeft(n))
 
   /** (Signed) rightshift of BigInt
    */
-  def >> (n: Int): BigInt = new BigInt(this.bigInteger.shiftRight(n))
-
+  def >> (n: Int): BigInt =
+    if (longEncoding && n >= 0) {
+      if (n < 64) new BigInt(_long >> n)
+      else if (_long < 0) BigInt(-1)
+      else BigInt(0) // for _long >= 0
+    } else new BigInt(this.bigInteger.shiftRight(n))
+  
   /** Bitwise and of BigInts
    */
-  def &  (that: BigInt): BigInt = new BigInt(this.bigInteger.and(that.bigInteger))
+  def &  (that: BigInt): BigInt =
+    if (this.longEncoding && that.longEncoding)
+      new BigInt(this._long & that._long)
+    else new BigInt(this.bigInteger.and(that.bigInteger))
 
   /** Bitwise or of BigInts
    */
-  def |  (that: BigInt): BigInt = new BigInt(this.bigInteger.or (that.bigInteger))
+  def |  (that: BigInt): BigInt =
+    if (this.longEncoding && that.longEncoding)
+      new BigInt(this._long | that._long)
+    else new BigInt(this.bigInteger.or (that.bigInteger))
 
   /** Bitwise exclusive-or of BigInts
    */
-  def ^  (that: BigInt): BigInt = new BigInt(this.bigInteger.xor(that.bigInteger))
+  def ^  (that: BigInt): BigInt =
+    if (this.longEncoding && that.longEncoding)
+      new BigInt(this._long ^ that._long)
+    else new BigInt(this.bigInteger.xor(that.bigInteger))
 
   /** Bitwise and-not of BigInts. Returns a BigInt whose value is (this & ~that).
    */
-  def &~ (that: BigInt): BigInt = new BigInt(this.bigInteger.andNot(that.bigInteger))
+  def &~ (that: BigInt): BigInt =
+    if (this.longEncoding && that.longEncoding)
+      new BigInt(this._long & ~that._long)
+    else new BigInt(this.bigInteger.andNot(that.bigInteger))
 
   /** Returns the greatest common divisor of abs(this) and abs(that)
    */
-  def gcd (that: BigInt): BigInt = new BigInt(this.bigInteger.gcd(that.bigInteger))
+  def gcd (that: BigInt): BigInt =
+    if (this.longEncoding) {
+      if (this._long == 0) return that.abs
+      // if (this._long == Long.MinValue) return (-this) gcd that
+      // this != 0 && this != Long.MinValue
+      if (that.longEncoding) {
+        if (that._long == 0) return this.abs
+        // if (that._long == Long.MinValue) return this gcd (-that)
+        new BigInt(BigInt.longGcd(this._long.abs, that._long.abs))
```

Comment:
why not check the cache here? gcd is often low, seems like we could often get lucky and find things in the cache.